File: tools/fashion-match-quality/retrieval/fashionClipAdapter.py
"""FashionCLIP embedding subprocess (Build 35 Workstream 03).

Real inference code: loads patrickjohncyh/fashion-clip via `transformers`
and embeds either images or text, L2-normalized, writing a JSON result to
stdout. This file is invoked by fashionClipAdapter.js exactly the way
tools/fashion-match-quality/l1/runL1.js shells out to Deno for the L1
harness - a pinned external runtime running the real model, not a
reimplementation of it in JS.

Protocol: argv[1] is a path to a request JSON file:
  {"mode": "image", "inputs": ["/abs/path/one.png", ...]}
  {"mode": "text",  "inputs": ["burgundy leather bomber jacket", ...]}
Writes one JSON object to stdout:
  {"ok": true, "modelRevision": "<resolved commit sha>", "embeddings": [[...], ...]}
  {"ok": false, "blocker": "...", "detail": "..."}
on any failure (never a bare traceback - the Node wrapper needs a parseable
result either way).
"""

import json
import sys


def fail(blocker, detail):
    print(json.dumps({"ok": False, "blocker": blocker, "detail": str(detail)[:2000]}))
    sys.exit(0)


def main():
    if len(sys.argv) != 2:
        fail("BAD_ARGS", "usage: fashionClipAdapter.py <request.json>")
        return

    try:
        with open(sys.argv[1], "r", encoding="utf-8") as f:
            request = json.load(f)
    except Exception as exc:  # noqa: BLE001 - this is a leaf process boundary
        fail("REQUEST_UNREADABLE", exc)
        return

    mode = request.get("mode")
    inputs = request.get("inputs") or []
    if mode not in ("image", "text") or not inputs:
        fail("BAD_REQUEST", f"mode must be 'image' or 'text' with a non-empty inputs list, got {request}")
        return

    try:
        import torch
        from transformers import CLIPModel, CLIPProcessor
    except Exception as exc:  # noqa: BLE001
        fail("RUNTIME_UNAVAILABLE", f"torch/transformers import failed: {exc}")
        return

    model_id = "patrickjohncyh/fashion-clip"
    try:
        model = CLIPModel.from_pretrained(model_id)
        processor = CLIPProcessor.from_pretrained(model_id)
    except Exception as exc:  # noqa: BLE001
        # Expected failure mode in any sandbox whose egress policy denies
        # huggingface.co and has no pre-seeded local cache - see
        # modelManifest.js's header comment.
        fail("MODEL_LOAD_FAILED", f"from_pretrained({model_id!r}) failed: {exc}")
        return

    revision = getattr(model.config, "_commit_hash", None) or "unknown"

    try:
        model.eval()
        with torch.no_grad():
            if mode == "image":
                from PIL import Image

                images = [Image.open(p).convert("RGB") for p in inputs]
                batch = processor(images=images, return_tensors="pt")
                features = model.get_image_features(**batch)
            else:
                batch = processor(text=inputs, return_tensors="pt", padding=True, truncation=True)
                features = model.get_text_features(**batch)

            features = features / features.norm(p=2, dim=-1, keepdim=True)
            embeddings = features.cpu().tolist()
    except Exception as exc:  # noqa: BLE001
        fail("INFERENCE_FAILED", exc)
        return

    print(json.dumps({"ok": True, "modelRevision": revision, "embeddings": embeddings}))


if __name__ == "__main__":
    main()

File: tools/fashion-match-quality/retrieval/fashionClipAdapter.js
'use strict';

/**
 * Node-side wrapper around the real FashionCLIP inference subprocess
 * (fashionClipAdapter.py). Mirrors tools/fashion-match-quality/l1/runL1.js's
 * pattern exactly: shell out to a pinned external runtime running the real
 * model, probe availability up front, and return `{ ok: false, blocker }`
 * rather than throwing when that runtime cannot run here - one blocked
 * embedding must never abort a whole evaluation pass.
 *
 * Why this is genuinely gated in this R&D session, not just defensively
 * coded: `isFashionClipAvailable()` checks three independent things, and any
 * one missing is real:
 *   1. `python3` is on PATH.
 *   2. `python3 -c "import torch, transformers"` succeeds (neither package is
 *      installed in this sandbox as of this build).
 *   3. A local HF cache for patrickjohncyh/fashion-clip already exists -
 *      `from_pretrained()` would otherwise need to reach huggingface.co,
 *      which this session's egress proxy explicitly policy-denies (see
 *      modelManifest.js's header). Checking this up front avoids spawning a
 *      subprocess that can only fail after a network timeout.
 * The moment all three hold in some environment, this module runs the real
 * model unmodified - nothing here needs to change.
 */

const { spawnSync } = require('node:child_process');
const fs = require('node:fs');
const os = require('node:os');
const path = require('node:path');

const ADAPTER_SCRIPT = path.join(__dirname, 'fashionClipAdapter.py');
const MODEL_ID = 'patrickjohncyh/fashion-clip';

function hfCacheDirFor(modelId) {
  // Standard huggingface_hub cache layout: models--<org>--<name>.
  const slug = `models--${modelId.replace('/', '--')}`;
  const base = process.env.HF_HOME || path.join(os.homedir(), '.cache', 'huggingface');
  return path.join(base, 'hub', slug);
}

let _availability;
/**
 * Returns { available, reasons } - reasons lists every unmet precondition,
 * never just the first, so a diagnostic report can show the whole picture
 * at once (spec section 4 asks this to be recorded, not just asserted).
 */
function checkFashionClipAvailability({ forceRecheck = false } = {}) {
  if (_availability && !forceRecheck) return _availability;

  const reasons = [];

  const pythonProbe = spawnSync('python3', ['--version'], { encoding: 'utf8' });
  const hasPython = pythonProbe.status === 0;
  if (!hasPython) reasons.push('PYTHON3_NOT_FOUND');

  let hasRuntime = false;
  if (hasPython) {
    const runtimeProbe = spawnSync('python3', ['-c', 'import torch, transformers'], { encoding: 'utf8' });
    hasRuntime = runtimeProbe.status === 0;
    if (!hasRuntime) {
      reasons.push('TORCH_OR_TRANSFORMERS_NOT_INSTALLED');
    }
  }

  const cacheDir = hfCacheDirFor(MODEL_ID);
  const hasLocalModelCache = fs.existsSync(cacheDir) && fs.readdirSync(cacheDir).length > 0;
  if (!hasLocalModelCache) {
    reasons.push('MODEL_NOT_CACHED_LOCALLY_AND_HUGGINGFACE_CO_IS_NETWORK_BLOCKED_IN_THIS_SESSION');
  }

  _availability = { available: reasons.length === 0, reasons, cacheDirChecked: cacheDir };
  return _availability;
}

function isFashionClipAvailable() {
  return checkFashionClipAvailability().available;
}

/**
 * Read the commit SHA of the locally-cached model snapshot, synchronously
 * and without a network call - huggingface_hub's on-disk cache layout keeps
 * `<cacheDir>/refs/<ref>` as a plain-text file containing the commit SHA a
 * ref resolved to at download time, and `<cacheDir>/snapshots/<sha>/` as the
 * actual snapshot directory. This lets a cache key be computed BEFORE
 * running the model (avoiding "the revision is only known after inference"
 * chicken/egg problem) whenever a real cached model is present. Returns
 * null when no cache exists (nothing to resolve).
 */
function resolveLocalCachedRevision() {
  const cacheDir = hfCacheDirFor(MODEL_ID);
  const refFile = path.join(cacheDir, 'refs', 'main');
  if (fs.existsSync(refFile)) {
    const sha = fs.readFileSync(refFile, 'utf8').trim();
    if (sha) return sha;
  }
  const snapshotsDir = path.join(cacheDir, 'snapshots');
  if (fs.existsSync(snapshotsDir)) {
    const entries = fs.readdirSync(snapshotsDir).sort();
    if (entries.length > 0) return entries[0];
  }
  return null;
}

/**
 * Embed a batch of inputs through the real FashionCLIP model.
 *
 * @param {'image'|'text'} mode
 * @param {string[]} inputs - absolute image file paths (mode 'image') or raw
 *   text strings (mode 'text').
 * @returns {{ok:true, modelRevision:string, embeddings:number[][]} |
 *           {ok:false, blocker:string, detail:string}}
 */
function embedBatch(mode, inputs) {
  const availability = checkFashionClipAvailability();
  if (!availability.available) {
    return {
      ok: false,
      blocker: 'FASHIONCLIP_UNAVAILABLE',
      detail: `FashionCLIP cannot run in this environment: ${availability.reasons.join(', ')}`,
    };
  }

  const tmpFile = path.join(os.tmpdir(), `fashionclip-req-${mode}-${process.pid}-${Date.now()}.json`);
  try {
    fs.writeFileSync(tmpFile, JSON.stringify({ mode, inputs }), 'utf8');

    const result = spawnSync('python3', [ADAPTER_SCRIPT, tmpFile], {
      encoding: 'utf8',
      maxBuffer: 64 * 1024 * 1024,
      timeout: 120000,
    });

    if (result.error) {
      return { ok: false, blocker: 'FASHIONCLIP_SUBPROCESS_FAILED', detail: result.error.message };
    }
    if (result.status !== 0) {
      return {
        ok: false,
        blocker: 'FASHIONCLIP_SUBPROCESS_FAILED',
        detail: (result.stderr || 'unknown python failure').slice(0, 2000),
      };
    }

    const parsed = JSON.parse(result.stdout.trim());
    return parsed;
  } catch (err) {
    return { ok: false, blocker: 'FASHIONCLIP_WRAPPER_EXCEPTION', detail: err.message };
  } finally {
    try {
      fs.unlinkSync(tmpFile);
    } catch {
      // best-effort cleanup
    }
  }
}

function embedImages(imagePaths) {
  return embedBatch('image', imagePaths);
}

function embedTexts(texts) {
  return embedBatch('text', texts);
}

module.exports = {
  MODEL_ID,
  isFashionClipAvailable,
  checkFashionClipAvailability,
  resolveLocalCachedRevision,
  embedImages,
  embedTexts,
};
